### PR TITLE
[SPIKE] content macros

### PR DESCRIPTION
**Pull request by petrjasek**
initial implementation.. let me know if it makes sense.
- [x] server side macro registration
- [x] `/macros` api
- [x] client widget macros
- [x] shortcuts for macros on client - it's `ctrl + macro.shortcut`

### CODE REVIEW

File: server/apps/macros.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+class MacrosResource(superdesk.Resource):
+    resource_methods = ['GET', 'POST']
+    item_methods = []
+    privileges = {'POST': 'archive'}
+
+    schema = {
+        'macro': {
+            'type': 'string',
+            'required': True,
+            'allowed': (macro['name'] for macro in macros)
```
Comment (amagdas) [inline server/apps/macros.py]: I don't think this really works as the macros are registered after the linking so here `macros` will be empty.
Probably doing a behave test with a wrong macro name will show this.

**Comment by amagdas** [comment]
Good job :heart_decoration: :100:

File: server/apps/macros.py (hunk 1 above)
Comment (petrjasek) [inline server/apps/macros.py]: that's why it's using generator ;) it's only evaluated when there is a post call

Comment (amagdas) [inline server/apps/macros.py]: That I get :) Still the schema is registered when the Eve app is created, so this lazy evaluation I don't think that it has the effect it should.

Comment (petrjasek) [inline server/apps/macros.py]: it's registered but until there is some iteration over allowed it's only passing references around.

Comment (amagdas) [inline server/apps/macros.py]: If that's the case, I learned something today :+1:

File: server/features/macros.feature [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            {"item": {"body_html": "$12"}}
+            """
+
+    @auth
+    Scenario: Return an error when triggering unknown macro
+        When we post to "/macros"
+            """
+            {"macro": "this does not exist!", "item": {"body_html": "test"}}
+            """
+        Then we get response code 400
```
Comment (amagdas) [inline server/features/macros.feature]: What's the error message?

File: server/settings.py [hunk 3]
```diff
@@ -143,7 +143,10 @@ def env(variable, fallback_value=None):
     'apps.packages',
     'apps.privilege',
     'apps.rule_sets',
-    'apps.highlights'
+    'apps.highlights',
+    'apps.macros',
+
+    'macros.currency',
```
Comment (akintolga) [inline server/settings.py]: It looks like we need to restart the server when there's a new macro, is that right?

**Comment by ioanpocol** [comment]
:+1:

File: server/settings.py (hunk 3 above)
Comment (amagdas) [inline server/settings.py]: We can load new modules dynamically if needed.

Comment (petrjasek) [inline server/settings.py]: yep we can instead specify a folder from where it would autoload all macros

Comment (akintolga) [inline server/settings.py]: I think it needs to be dynamic. So I guess settings.py doesn't need to know about the individual modules.

**Comment by amagdas** [comment]
Rebase needed...

**Comment by akintolga** [comment]
Happy :+1: 

I have updated the wiki page with what we could gather from Michael. I added the "Additional Requirements" section. I think (maybe) we can start using the current implementation and then we can attempt to address the remaining requirements in later milestones. This needs to be discussed with POs, PLs and ORPs (other responsible people :sweat_smile: ).

**Pull request merged**